File: slp/backend-dotnet/Features/User/UserRepository.cs
using backend_dotnet.Data;
using backend_dotnet.Features.Dashboard;
using backend_dotnet.Features.User;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.User;

public class UserRepository : IUserRepository
{
    private readonly AppDbContext _db;

    public UserRepository(AppDbContext db)
    {
        _db = db;
    }

    public async Task<User?> GetByIdAsync(int id)   // matches interface
    {
        return await _db.Users.FindAsync(id);
    }

    public async Task<User?> GetByEmailAsync(string email)
    {
        return await _db.Users.FirstOrDefaultAsync(x => x.Email == email);
    }

    public async Task<User?> GetByUsernameAsync(string username)
    {
        return await _db.Users.FirstOrDefaultAsync(x => x.Username == username);
    }

    public async Task<User?> GetByResetTokenAsync(string token)
    {
        return await _db.Users.FirstOrDefaultAsync(x => x.PasswordResetToken == token);
    }

    public async Task<User?> GetByEmailVerificationTokenAsync(string token)
    {
        return await _db.Users.FirstOrDefaultAsync(x => x.EmailVerificationToken == token);
    }

    public async Task UpdateAsync(User user)
    {
        _db.Users.Update(user);
        await _db.SaveChangesAsync();
    }

    public async Task CreateAsync(User user)
    {
        _db.Users.Add(user);
        await _db.SaveChangesAsync();
    }

    public async Task DeleteAsync(User user)
    {
        _db.Users.Remove(user);
        await _db.SaveChangesAsync();
    }

    public async Task<IEnumerable<User>> GetAllAsync(string? search = null)
    {
        var query = _db.Users.AsQueryable();
        if (!string.IsNullOrWhiteSpace(search))
        {
            var pattern = $"%{search}%";
            query = query.Where(u =>
                EF.Functions.ILike(u.Username, pattern) ||
                EF.Functions.ILike(u.Email, pattern));
        }
        return await query.OrderBy(u => u.Id).ToListAsync();
    }

    public async Task<UserStatsDto> GetUserStatsAsync(int userId)
    {
        var quizCount = await _db.Quizzes
            .Where(q => q.UserId == userId && !q.Disabled)
            .CountAsync();

        var questionCount = await _db.Questions
            .Where(q => q.UserId == userId)
            .CountAsync();

        var sourceCount = await _db.Sources
            .Where(s => s.UserId == userId && s.DeletedAt == null)
            .CountAsync();

        var favoriteCount = await _db.FavoriteItems
            .Where(f => f.UserId == userId)
            .CountAsync();

        return new UserStatsDto
        {
            QuizCount = quizCount,
            QuestionCount = questionCount,
            SourceCount = sourceCount,
            FavoriteCount = favoriteCount
        };
    }
}

File: slp/backend-java/src/main/java/com/app/backendjava/features/user/UserRepository.java
package com.app.backendjava.features.user;

import com.app.backendjava.features.dashboard.UserStatsDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface UserRepository extends JpaRepository<User, Integer>, IUserRepository {

    @Override
    default Optional<User> getByIdAsync(int id) { return findById(id); }

    @Override
    @Query("SELECT u FROM User u WHERE u.email = :email")
    Optional<User> getByEmailAsync(@Param("email") String email);

    @Override
    @Query("SELECT u FROM User u WHERE u.username = :username")
    Optional<User> getByUsernameAsync(@Param("username") String username);

    @Override
    @Query("SELECT u FROM User u WHERE u.passwordResetToken = :token")
    Optional<User> getByResetTokenAsync(@Param("token") String token);

    @Override
    @Query("SELECT u FROM User u WHERE u.emailVerificationToken = :token")
    Optional<User> getByEmailVerificationTokenAsync(@Param("token") String token);

    @Override
    default User saveAsync(User user) { return save(user); }

    @Override
    default void deleteAsync(User user) { delete(user); }

    @Override
    @Query("SELECT u FROM User u WHERE (:search IS NULL OR " +
            "LOWER(u.username) LIKE LOWER(CONCAT('%', :search, '%')) OR " +
            "LOWER(u.email) LIKE LOWER(CONCAT('%', :search, '%')))")
    Page<User> getAllAsync(@Param("search") String search, Pageable pageable);

    // Note: This requires Quizzes, Questions, Sources, and FavoriteItems entities to exist
    @Query("SELECT new com.app.backendjava.features.dashboard.UserStatsDto(" +
            "(SELECT COUNT(q) FROM Quiz q WHERE q.userId = :userId AND q.disabled = false), " +
            "(SELECT COUNT(qn) FROM Question qn WHERE qn.userId = :userId), " +
            "(SELECT COUNT(s) FROM Source s WHERE s.userId = :userId AND s.deletedAt IS NULL), " +
            "(SELECT COUNT(f) FROM FavoriteItem f WHERE f.userId = :userId))")
    UserStatsDto getUserStatsAsync(@Param("userId") int userId);
}